src/ingameWindows/iwMilitaryBuilding.cpp

```diff
@@ -106,14 +106,30 @@ void iwMilitaryBuilding::Msg_PaintAfter()
                 (*it)->GetGOT() == GOT_NOF_DEFENDER ||
                 (*it)->GetGOT() == GOT_NOF_PASSIVESOLDIER)
         {
-            soldiers.insert(static_cast<nofSoldier*>(*it)->GetRank());
+            soldiers.insert(static_cast<nofSoldier*>(*it));
         }
     }
 
     // Soldaten zeichnen
     unsigned short i = 0;
-    for(std::multiset<unsigned>::iterator it = soldiers.begin(); it != soldiers.end(); ++it, ++i)
-        LOADER.GetMapImageN(2321 + *it)->Draw(GetX() + width_ / 2 - 22 * TROOPS_COUNT[building->nation][building->size] / 2 + 12 + i * 22, GetY() + 110, 0, 0, 0, 0, 0, 0);
+    for(std::multiset<nofSoldier*>::const_iterator it = soldiers.begin(); it != soldiers.end(); ++it, ++i)
+        LOADER.GetMapImageN(2321 + (*it)->GetRank())->Draw(GetX() + width_ / 2 - 22 * TROOPS_COUNT[building->nation][building->size] / 2 + 12 + i * 22, GetY() + 110, 0, 0, 0, 0, 0, 0);
+
+    // Draw health above soldiers
+    if (GAMECLIENT.GetGGS().isEnabled(ADDON_MILITARY_HITPOINTS)) { 
+		unsigned short leftXCoordinate = GetX() + width_ / 2 - 22 * TROOPS_COUNT[building->nation][building->size] / 2;
+
+		// black background for hitpoints
+		DrawRectangle(leftXCoordinate , GetY() + 84, 22 * TROOPS_COUNT[building->nation][building->size], 14, 0x96000000);
+
+        i = 0;
+        for (std::multiset<nofSoldier*>::const_iterator it = soldiers.begin(); it != soldiers.end(); ++it, ++i) {
+            char txt[64];
+            sprintf(txt, "%u/%u", (*it)->GetHitpoints() , HITPOINTS[building->nation][(*it)->GetRank()]);
```

Comment:
Can you please use a std::string for that? There is a toString method in the helpers folder for converting the numbers to a string or use a stringstream. sprintf is unsafe (although used correctly in this case)